Add generic preprocessors

TuringVerified opened the pull request:
For now, I've added rules to filter out IPv4, IPv6 address & URLs starting with http/https in the ICU file. Most of regex_replace.py is the same as split_japanese_phrases.py.

I haven't added any tests yet. I was working using the code for the preprocessor test suite and when using QueryConfig inside a test file, the ICU-related settings (such as replacements) do not seem to be loaded. Is that supposed to happen? Are there any functions that can allow me to load the ICU file? Or is this something I'd have to make myself.

lonvia commented on settings/icu_tokenizer.yaml:
```diff
@@ -1,5 +1,13 @@
 query-preprocessing:
     - step: split_japanese_phrases
+    - step: regex_replace
```
As said in the issue, please don't add anything here in the default config. You can use the IPv4/6 example in the tests and in the [documentation](https://github.com/osm-search/Nominatim/blob/master/docs/customize/Tokenizers.md#query-preprocessing) that still needs to be written.

lonvia replied:
The code is still here.

lonvia commented on src/nominatim_api/query_preprocessing/regex_replace.py:
```diff
@@ -0,0 +1,46 @@
+    def __call__(self, phrases: List[Phrase]) -> List[Phrase]:
+        """Apply regex replacements to the given addresses.
+        """
+        return [self.split_phrase(p) for p in phrases]
```
This is not yet implementing the requirement that phrases are dropped completely when they are empty.

lonvia commented on src/nominatim_api/query_preprocessing/regex_replace.py:
```diff
@@ -0,0 +1,46 @@
+
+        match_patterns = self.config.get('replacements', 'Key not found')
+        for item in match_patterns:
+            phrase.text = re.sub(item['pattern'], item['replace'], phrase.text)
```
This can be sped up by pre-compiling the regular expression.

lonvia reviewed (commented):
Regarding the tests: these are supposed to work independently from the default configuration. Simply initialize the QueryConfig with a dictionary containing the configuration (e.g. `{'step': 'regex_replace', 'replacements', ...}`)

When writing the tests, do not only think about successful use cases but also about possible errors. Please also test that the different variants of backreferences work as expected. See the [Python documentation](https://docs.python.org/3/library/re.html#re.sub) for more info.

TuringVerified commented:
I implemented the requested changes

lonvia commented on docs/customize/Tokenizers.md:
```diff
@@ -106,6 +111,17 @@ The following is a list of preprocessors that are shipped with Nominatim.
+    replacements:
+        - pattern: regex pattern
+          replace: string to replace with
+
```
Have you compiled the documentation and looked at the result of your changes?

lonvia replied:
The documentation is built using mkdocs, see the [Developer's documentation](https://nominatim.org/release-docs/develop/develop/Development-Environment/#documentation-pages). This page also has a list of packages you need to install to make it work.

lonvia commented on test/python/api/query_processing/test_regex_replace.py:
```diff
@@ -0,0 +1,48 @@
+
+    out = run_preprocessor_on(query)
+
+    assert out == [qmod.Phrase(qmod.PHRASE_ANY, outp)]
```
This test clearly shows that the output if the preprocessors contains Phrases with the empty string.

TuringVerified replied:
I was a bit confused about what "dropping phrases when they're empty" meant. Do I just return a list if the end output is empty?

lonvia replied:
Yes, if there was just one phrase, then the list should be empty in the end. If there are other Phrases, then they need to be kept, of course.

TuringVerified commented:
I got a bit confused due to the symlinks and thought that the config file was updated, but it wasn't. Fixed that now.

I compiled the documentation and fixed the indentation issue. Also, when there's a singular empty phrase, we should now just get an empty list straight. I've updated the tests to reflect this too.

TuringVerified commented:
I added docstrings in the .py file and I set members to True here so that it should all visible on the page at once now. But that also shows private functions. Is that something I should include? Or do I just add docstrings in the Python file itself?

lonvia commented:
You don't need the `members: True`. The documentation is taken from the module documentation, i.e. the docstring [here](https://github.com/osm-search/Nominatim/pull/3675/files#diff-a3919314ceca21964ea312baa1dfb83627556796d8f325e0363b0280af83e2a6R8).

TuringVerified commented:
Hi, I realised that I have a merge in my commit history and it seems to be complicating rebasing. For this PR, would you prefer a single squashed commit for simplicity? Or should I preserve the commit history?

lonvia commented:
The history should not have any merge commits. `git fetch origin; git rebase origin/master` usually does the trick.

Apart from that, it is your choice if you want to squash all into one commit or split up in multiple commits. It would be nice if the commits are somewhat meaningful. If there are "fixup" commits, squash them together with the commit they are fixing. (It takes a bit of getting used to but `git rebase -i master` is a really good tool to sort your commits.)

Feel free to force-push to the PR branch any time after rebasing.

lonvia reviewed (approved):
Rebase worked like a charm. Thank you.
